biz.aQute.launcher/src/aQute/launcher/plugin/ProjectLauncherImpl.java

```diff
@@ -205,8 +207,7 @@ public void prepare() throws Exception {
 			return;
 		prepared = true;
 		super.prepare();
-
-		writeProperties();
+		calculatedProperties(getRunProperties());
```

Comment:
`getRunProperties()` returns by-reference, and `calculatedProperties()` modifies it in-place. Is this what we want? I noticed that `RemoteProjectLauncherPlugin` makes a copy of `runproperties` before passing it in to `calculatedProperties()`.

Reply:
I also noted that there is no code path in which `calculatedProperties()` is called from within an `update()` call. Is this by design, or an oversight?

Reply:
`getRunProperties()` is fine since `updateFromProject()` sets the `runproperties` from `getProject().getRunProperties()` which returns a new map. `RemoteProjectLauncherPlugin` is somewhat of a mystery to me, so I don't know why it does what it does. 🤷‍♂️ 



> I also noted that there is no code path in which `calculatedProperties()` is called from within an `update()` call. Is this by design, or an oversight?

Mistake on my part. I'll push an updated commit. Good catch!